### ISSUE TITLE
Unexpected result for distinct() on tbl_dt

### ISSUE THREAD

**Issue by BillyRen** [hadley/dplyr#1208]
**The result of `distinct()` is different from `distinct(all col names)` on an `tbl_dt` object**.

For instance, `data` is an `tbl_dt` object with ten columns: col1,.., col10.

The result of `data %>% distinct()` is different from `data %>% distinct(col1, ..., col10)`.

**Comment by hadley** [comment]
Could you please provide a minimal reproducible example?

**Comment by pimentel** [comment]
Could it be related to this PR?

https://github.com/hadley/dplyr/pull/990

**Comment by BillyRen** [comment]
Hi pimentel,

I think that's the problem.

I generated the tbl_dt data from this step:
    
    fread(filepath) %>%
        tbl_dt() %>%
        left_join(another tbl_dt)

Now, the `tbl_dt` object has a key, **BUT I DO NOT SET IT**,
it seems that it **automatically set the key** when run the `left join` process.

Then the `distinct()` method  only returns the unique key column.

At #990 , you said this change has been merged into the master branch.
Should I use the github version rather than the cran version to fix this?

**Comment by pimentel** [comment]
Yup, a join on a `tbl_dt` will set a key (required by `data.table` on joins).

It's fixed in my local version, but that is now an old development version so I wouldn't suggest installing my version. Perhaps @hadley can merge it then you can use the current development version?

Alternatively you can fork the current development version and add the minor change (it's just one line, see the diff)

**Comment by romainfrancois** [comment]
Same problem as #1285 I guess

**Comment by dbuijs** [comment]
Confirm this still happens on 0.4.3. Happy to see it flagged and looking forward to 0.5.
Workaround right now is to just explicitly specify all columns on any tbl_dt. I suppose we could also just set a new key with all the columns.
